Variant Constants

bonjorno7 opened the pull request:
Here's my attempt at variant type constants which can be added to any class.
It messes with a lot of files and will need thorough review from people much smarter than I.

You probably want to know what problem this solves.
I was refactoring Color conversions and constants into a module outside of core, and found out that it isn't possible to add non-integer constants.

Here's the related discussion: https://github.com/godotengine/godot-proposals/discussions/6925

Things I still need to look into:
- C# (did some things but still doesn't work)
- Colored squares to the right of color constants
- ~~Whatever the automated tests complain about~~

bonjorno7 committed:
Attempt at adding variant constants

Not working yet.

bonjorno7 committed:
Make variant constants worse

This actually breaks it even further.

bonjorno7 committed:
Fix variant constant calling wrong function

Huh, it was closer to working than I thought it was.
Fixing this actually makes variant constants work.
Still need to figure out how to make it put the right type in the documentation; right now it's just `(1, 0, 0, 1)` where it should be `Color(1, 0, 0, 1)`.
It does behave like a Color in GDScript though so that's good.

bonjorno7 committed:
get_variant_constant return null instead of 0

This makes more sense as a default return value because this function can return any variant type, not just integers.

bonjorno7 committed:
Add documentation for variant constants to ClassDB

Pretty much just copy paste from the integer constant functions with different keywords.

bonjorno7 committed:
Use get_construct_string instead of stringify

Now the documentations knows what type it is.
Not sure how it'd behave for NIL, but I guess it doesn't matter because you'd never make a null constant.

bonjorno7 committed:
Fix KeyValue type for get_variant_constant_list

Classic copy and paste error.

bonjorno7 committed:
Add variant constants to get_api_hash

I have no idea what this code does, but I'm pretty sure it's important.

bonjorno7 commented:
Should I open a proposal or a discussion for this (over on [godot-proposals](https://github.com/godotengine/godot-proposals))?

AThousandShips commented:
Yes please do, this is somewhat significant and has many details so discussion and analysis of how it should best be done is important

bonjorno7 commented:
https://github.com/godotengine/godot-proposals/discussions/6925
I don't really see the point of a proposal if there's already a pull request, but a discussion is useful I think.

AThousandShips commented:
Part of the reason is that the PR should be primarily for the discussion of implementation, not the how and why when it is more general or abstract

Also it's good to have if this PR for some reason or another doesn't get merged and avoiding any discussion to be hidden in a closed PR

bonjorno7 commented:
Fair enough.
I still think a discussion is better than an issue though, and it isn't necessary to have both.

YuriSizov commented:
You should open a proposal issue. A discussion is a place to have a free form conversation about a problem or a potential feature, but you already have a technical implementation. A formal proposal is required before a feature can be accepted, and it should explain why this is needed and why this implementation is the correct one. It's the right place to gauge the need from other users too.

bonjorno7 commented:
Alright then here: https://github.com/godotengine/godot-proposals/issues/6926

bonjorno7 committed:
Load variant constants into the correct list

This is why I don't like such short variable names, but I'm not changing it either cause I'm trying to match the surrounding code.

bonjorno7 committed:
Add variant constant support to GDExtension

No idea if it works or how to test it but it's probably fine.

bonjorno7 committed:
Add variant constant support to Mono

I have my doubts whether this will work.
Haven't looked into global constants at all yet either.
In class_db_api_to_json the constants are sorted alphabetically, as are the variant constants, but they are sorted separately and appended.

bonjorno7 committed:
Fix appending to constants array

Can't append array because it's a Variant, so I'm doing it like this now.

bonjorno7 committed:
Use class_doc.constants instead of variant_constants

Just remembered I'm reusing the constants list because the type is String so why not.

fire commented:
aaronfranke was interested in using floats for constants, and this would allow that.

Can you provide a review? @aaronfranke

This is not merg-able as is due to there being 13 commits.

https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html

bonjorno7 commented:
It's not squashed yet because it's still quite a mess.
Ideally someone who knows the engine better takes my changes and builds on them for a proper PR.
